### PR TITLE
Fix not retryable error codes from: Unavailable -> FailedPrecondition

### PR DESCRIPTION
**Pull request by ptabor**
- ErrGRPCNotCapable("etcdserver: not capable") -> codes.FailedPrecondition  (it will not autofix, it requires new version of server)
 - ErrGPRCNotSupportedForLearner("etcdserver: rpc not supported for learner") -> codes.FailedPrecondition (as long as its learner, the call will not work)
 - ErrGRPCClusterVersionUnavailable("etcdserver: cluster version not found during downgrade") -> codes.FailedPrecondition (backend does not contain the version (old etcd?) so retry will not help)

https://github.com/etcd-io/etcd/runs/2599598633?check_suite_focus=true

```
{"level":"warn","ts":"2021-05-17T09:55:30.246Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc000539880/#initially=[unix://localhost:m30]","attempt":0,"error":"rpc error: code = Unavailable desc = etcdserver: rpc not supported for learner"}
{"level":"warn","ts":"2021-05-17T09:55:30.270Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying
of unary invoker
failed","target":"etcd-endpoints://0xc000539880/#initially=[unix://localhost:m30]","attempt":1,"error":"rpc
error: code = Unavailable desc = etcdserver: rpc not supported for
learner"}`
```

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by ptabor** [comment]
Seems its not as obvious as it seems: 
 
 `TestBalancerSupportLearner` tests that requests (not supported by learners) send to learners are getting retried. This makes sense if the retry happens on another instance that is likely not a learner. 
But if the learner(s) the only endpoint that client connects to, the retrying does not make sense.

**Comment by xiang90** [comment]
LGTM.

Even for the learner case, this is not a transient error from the server point of view. So the client should not rely on `Unavailable` to trigger the rebalance to the actual etcd server.

**Review by gyuho — APPROVED** [review]
lgtm, thanks for the fix

**Comment by ptabor** [comment]
PTAL. Decided to add dedicated case to retry interceptor: 
  https://github.com/etcd-io/etcd/pull/12985/files#diff-4293a62cbd502b5fb268bfe85b304addfe3348237b9241f7b9218d1a0de5e237

File: api/v3rpc/rpctypes/error.go [hunk 1]
```diff
@@ -68,19 +68,19 @@ var (
...
-	ErrGRPCNotCapable                 = status.New(codes.Unavailable, "etcdserver: not capable").Err()
+	ErrGRPCNotCapable                 = status.New(codes.FailedPrecondition, "etcdserver: not capable").Err()
 	ErrGRPCStopped                    = status.New(codes.Unavailable, "etcdserver: server stopped").Err()
 	ErrGRPCTimeout                    = status.New(codes.Unavailable, "etcdserver: request timed out").Err()
 	ErrGRPCTimeoutDueToLeaderFail     = status.New(codes.Unavailable, "etcdserver: request timed out, possibly due to previous leader failure").Err()
 	ErrGRPCTimeoutDueToConnectionLost = status.New(codes.Unavailable, "etcdserver: request timed out, possibly due to connection lost").Err()
 	ErrGRPCUnhealthy                  = status.New(codes.Unavailable, "etcdserver: unhealthy cluster").Err()
 	ErrGRPCCorrupt                    = status.New(codes.DataLoss, "etcdserver: corrupt cluster").Err()
-	ErrGPRCNotSupportedForLearner     = status.New(codes.Unavailable, "etcdserver: rpc not supported for learner").Err()
+	ErrGPRCNotSupportedForLearner     = status.New(codes.InvalidArgument, "etcdserver: rpc not supported for learner").Err()
```
Comment (xiang90) [inline api/v3rpc/rpctypes/error.go:78]: It seems `Unimplemented` or `Not Found` status is better fit here?

File: client/v3/retry_interceptor.go [hunk 2]
```diff
@@ -297,17 +298,26 @@ func waitRetryBackoff(ctx context.Context, attempt uint, callOpts *options) erro
 }
 
 // isSafeRetry returns "true", if request is safe for retry with the given error.
-func isSafeRetry(lg *zap.Logger, err error, callOpts *options) bool {
+func isSafeRetry(c *Client, err error, callOpts *options) bool {
 	if isContextError(err) {
 		return false
 	}
+
+	// Situation when learner refuses RPC its supposed to not serve is from server
```
Comment (xiang90) [inline client/v3/retry_interceptor.go:306]: Yea. We should have a more reliable way for client to retry on the learner case. The client should be able to tell if the server instance is a leaner or not.

File: api/v3rpc/rpctypes/error.go (hunk 1 above)
Comment (ptabor) [reply api/v3rpc/rpctypes/error.go:78]: How about FailedPrecondition ? 
 - not found - sounds like a problem with 'resource' not 'semantic'. 
 - not implemented - sounds like a way to fix the problem is to upgrade the server. 
 
FailedPrecondition is the most neutral in my opinion

File: client/v3/retry_interceptor.go (hunk 2 above)
Comment (ptabor) [reply client/v3/retry_interceptor.go]: Added TODO

File: api/v3rpc/rpctypes/error.go (hunk 1 above)
Comment (xiang90) [reply api/v3rpc/rpctypes/error.go]: `Not implemented` can mean two things:

1. The operation is not implemented
2. The operation is not supported/enabled

This case can fall into the second one.

ref: https://grpc.github.io/grpc/core/md_doc_statuscodes.html.

The precondition failure indicates that the service is in a state that cannot execute the operation. Usually, it is an external triggered state and can be fixed externally (for example to remove an unempty director, client needs to remove all files under it). `learner` is an internal state of the service and cannot be changed by users, but it also makes sense in a border sense.

I have no strong opinion.

**Comment by xiang90** [comment]
LGTM

**Review by gyuho — APPROVED** [review]

**Comment by gyuho** [comment]
@ptabor This is a bug fix. Can we backport to 3.5?

**Pull request merged**